flink-staging/flink-scala-shell/src/main/scala/org/apache/flink/api/scala/FlinkShell.scala

```diff
@@ -61,99 +73,217 @@ object FlinkShell {
+            deployNewYarnCluster(yarnConfig)
+          case None => // there is no information for new cluster. Then we use yarn properties.
+            fetchDeployedYarnClusterInfo()
+        }
 
-          case None =>
-            new FlinkILoop(host, port, externalJars)
-        })
+      case ExecutionMode.UNDEFINED => // Wrong input
+        throw new IllegalArgumentException("please specify execution mode:\n" +
+          "[local | remote <host> <port> | yarn]")
+    }
+  }
 
-      val settings = new Settings()
+  def startShell(config: Config): Unit = {
+    println("Starting Flink Shell:")
 
-      settings.usejavacp.value = true
-      settings.Yreplsync.value = true
+    val (repl, cluster) = try {
+      val (host, port, cluster) = fetchConnectionInfo(config)
+      println(s"\nConnecting to Flink cluster (host: $host, port: $port).\n")
+      val repl: Option[FlinkILoop] = bufferedReader match {
+        case Some(reader) =>
+          val out = new StringWriter()
+          Some(new FlinkILoop(host, port, config.externalJars, reader, new JPrintWriter(out)))
+        case None =>
+          Some(new FlinkILoop(host, port, config.externalJars))
+      }
+
+      (repl, cluster)
+    } catch {
+      case e: IllegalArgumentException =>
+        println(s"Error: ${e.getMessage}")
+        sys.exit()
+    }
 
-      // start scala interpreter shell
+    val settings = new Settings()
+    settings.usejavacp.value = true
+    settings.Yreplsync.value = true
+
+    try {
       repl.foreach(_.process(settings))
     } finally {
       repl.foreach(_.closeInterpreter())
-      cluster.foreach(_.stop())
+      cluster match {
+        case Some(Left(miniCluster)) => miniCluster.stop()
+        case Some(Right(yarnCluster)) => yarnCluster.shutdown(false)
+        case _ =>
+      }
+    }
+
+    println(" good bye ..")
+  }
+
+  def deployNewYarnCluster(yarnConfig: YarnConfig) = {
+    val yarnClient = FlinkYarnSessionCli.getFlinkYarnClient
+
+    // use flink-dist.jar for scala shell
+    val jarPath = new Path("file://" +
+      s"${yarnClient.getClass.getProtectionDomain.getCodeSource.getLocation.getPath}")
+    yarnClient.setLocalJarPath(jarPath)
+
+    // load configuration
+    val confDirPath = CliFrontend.getConfigurationDirectoryFromEnv
+    val flinkConfiguration = GlobalConfiguration.getConfiguration
+    val confFile = new File(confDirPath + File.separator + "flink-conf.yaml")
+    val confPath = new Path(confFile.getAbsolutePath)
+    GlobalConfiguration.loadConfiguration(confDirPath)
+    yarnClient.setFlinkConfigurationObject(flinkConfiguration)
+    yarnClient.setConfigurationDirectory(confDirPath)
+    yarnClient.setConfigurationFilePath(confPath)
+
+    // number of task managers is required.
+    yarnConfig.containers match {
+      case Some(containers) => yarnClient.setTaskManagerCount(containers)
+      case None =>
+        throw new IllegalArgumentException("Number of taskmanagers must be specified.")
```

Comment:
Why is the number of yarn container an optional command line parameter if we actually require it? Shouldn't we fail much earlier then? E.g. when the user does not provide the number of containers.